RingBufCPP.h

```diff
@@ -70,14 +114,47 @@ bool pull(Type *dest)
     return ret;
 }
 
+/**
+* Remove "length" last elements from buffer, and copy them to "dest" buffer
+* Return: number of elements actually copied (could be less if numElements < length).
+*/
+size_t pull(Type *dest, size_t length)
+{
+    size_t tail;
+    size_t ret = 0;
+
+    RB_ATOMIC_START
+    {
+        if ((!isEmpty()) && (length > 0)) {
+            length = _numElements > length ? length : _numElements; // Trunk length if exceeds elements available for reading.
+
+            tail = getTail();
+            if(length > (MaxElements-tail)) {
+                // Needs two memcpy.
+                memcpy(dest,&_buf[tail],sizeof(Type)*(MaxElements-tail)); // From tail to the end of _buf
```

Comment:
memcpy does not call the copy constructor, this might cause things to break on things that rely on it.

Reply:
100% valid and I didn't think about it. For the application I was using your library I didn't needed the copy constructor. But it's a very valid point. I don't know if there's C++ equivalent for making the copy using the copy constructor. A for maybe? 

I want to avoid calling multiple times to the same function and calculating the head or the tail position each time. It's no optimal for a low powered microcontroller.

Reply:
What about if we change memcpy for std:copy?

Reply:
Yeah, std::copy_n would invoke the copy constructor. One thing to consider though is the std:: library might not be 
 available on embedded platforms. 

Generally these types of optimizations should be guided by profiling and/or benchmark data. Unless you have data showing this is a substantial speedup for certain use cases, I do not really see a compelling reason to add it. Keep in mind determinism and latency is generally more important on realtime embedded systems then performance is. This change causes interrupts to be masked for O(N) cycles, whereas the current implementation is O(1). If my system needed to be able to respond to a certain interrupt within a latency constraint, a change like this could dramatically impact this. Just trying to point out here that I agree you are likely to see some marginal performance increase, it is coming at a tradeoff.

Reply:
I undestand. Seeing all these comments, I think it's safe to skip this pull request. You are right about the latency of this modificaction. In my application that wasn't important because I was not using this library in anything interrupt related. I guess my implementation is valid if you want to use the ring buffer for something thread safe or without interrupts.

Thank anyway for taking the time to give me feedback on these changes.